lib/xverifier/applicator.rb

```diff
@@ -2,77 +2,181 @@
       end
     end
 
+    # ProcApplicatior is used when #to_proc is available.
+    # It works not just with procs, but also with hashes etc
+    # @example with proc
+    #   Applicator.call(-> { foo }, binding_, context) # => foo
+    # @example with hash
+    #   Applicator.call(Hash[foo: true], binding_, :foo) # => true
+    #   Applicator.call(Hash[foo: true], binding_, :bar) # => nil
     class ProcApplicatior < self
-      def call(binding, context)
-        invoke_lambda(applicable.to_proc, binding, context)
+      # @param applicable [#to_proc]
+      # @return ProcApplicatior if applicable accepts #to_proc
+      # @return [nil] otherwise
+      def self.build_class(applicable)
+        self if applicable.respond_to?(:to_proc)
+      end
+
+      # @param binding_ [#instance_exec] target to applicate applicable
+      # @param context additional info to send it to applicable
+      # @return application result
+      def call(binding_, context)
+        invoke_lambda(applicable.to_proc, binding_, context)
       end
     end
 
+    # Quoter is used when there is no other way to applicate applicatable.
+    # @example
+    #   Applicator.call(true, binding_, context) # => true
     class Quoter < self
+      # @return applicable without changes
       def call(*)
         applicable
       end
     end
+
+    extend ClassBuilder::Mixin
+
+    self.buildable_classes =
+      [Proxy, MethodExtractor, InstanceEvaluator, ProcApplicatior, Quoter]
+
+    attr_accessor :applicable
+
+    # Applicates applicable on binding_ with context
+    # @todo add @see #initialize when it's todo done
+    # @param applicable [applicable]
+    #   see examples in sublcasses defenitions
+    # @param binding_ [#instance_exec]
+    #   where should applicable be applied. It could be either a generic object,
+    #   where it would be `instance_exec`uted, or a binding_
+    # @param context
+    #   geneneric data you want to pass to applicable function.
+    #   If applicable would not accept params, it would not be sent
+    # @return application result
+    def self.call(applicable, binding_, context)
+      build(applicable).call(binding_, context)
+    end
+
+    # Always use build instead of new
+    # @todo add more examples right here
+    # @param applicable [applicable]
+    #   see examples in sublcasses defenitions
+    # @api private
+    def initialize(applicable)
+      self.applicable = applicable
+    end
+
+    # @param [Applicator] other
+    # @return [Boolean] true if applicable matches, false otherwise
+    def ==(other)
+      applicable == other.applicable
+    end
+
+    # @!method call(binding_, context)
+    #   @abstract
+    #   Applicates applicable on binding_ with context
+    #   @param binding_ [#instance_exec] biding would be used in application
+    #   @param context param would be passed if requested
+    #   @return application result
+
+    private
+
+    # invokes lambda basing on it's arity
```

Comment:
Doesn't it mean like "select some lambda"
Used: "invokes lambda respecting it's arity"